lib/command.js

```diff
@@ -3,14 +3,79 @@
 let Heroku = require('heroku-client')
 let cli = require('..')
 
-function apiMiddleware (response, cb) {
-  let warning = response.headers['x-heroku-warning'] || response.headers['warning-message']
-  if (warning) cli.action.warn(warning)
-  cb()
+function twoFactorWrapper (options, preauths, context) {
+  return function (res, buffer) {
+    let body
+    try {
+      body = this.parseBody(buffer)
+    } catch (e) {
+      this._handleFailure(res, buffer)
+    }
+
+    // safety check for if we have already seen this request for preauthing
+    // this prevents an infinite loop in case some preauth fails silently
+    // and we continue to get two_factor failures
+
+    // this might be better done with a timer in case a command takes too long
+    // and the preauthorization runs out, but that seemed unlikely
+    if (res.statusCode === 403 && body.id === 'two_factor' && !preauths.requests.includes(this)) {
+      twoFactorPrompt(options, context)
+      .then(function (secondFactor) {
+        // default preauth to always happen unless explicitly disabled
+        if (options.preauth === false) {
+          this.options.headers = Object.assign({}, this.options.headers, {'Heroku-Two-Factor-Code': secondFactor})
+          this.request()
+        } else {
+          preauths.requests.push(this)
+
+          // if multiple requests are run in parallel for the same app, we should
+          // only preauth for the first so save the fact we already preauthed
+          if (!preauths.promises[body.app.name]) {
+            preauths.promises[body.app.name] = cli.preauth(body.app.name, heroku(context), secondFactor)
+          }
+
+          preauths.promises[body.app.name].then(function () {
+            this.request()
+          }.bind(this))
```

Comment:
I think it might make sense to store `this` above in the scope so you don't have to keep binding:

```js
let self = this
```
```